src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs

```diff
@@ -133,115 +133,115 @@ internal static string FormatLiteral(bool value)
+        private static void FormatChar(
+            ref PooledStringBuilder pooledBuilder,
+            string str,
+            int index,
+            char c,
+            char quote,
+            bool useLanguageSpecificEscapes,
+            bool useUnicodeEscapes)
         {
-            PooledStringBuilder pooledBuilder = null;
-            StringBuilder sb = null;
-            int lastEscape = -1;
-            for (int i = 0; i < str.Length; i++)
-            {
-                char c = str[i];
-                char replaceWith = '\0';
-                bool unicodeEscape = false;
-                switch (c)
-                {
-                    case '\\':
-                        replaceWith = c;
-                        break;
-
-                    case '\0':
-                        replaceWith = '0';
-                        break;
-
-                    case '\r':
-                        replaceWith = 'r';
-                        break;
-
-                    case '\n':
-                        replaceWith = 'n';
-                        break;
-
-                    case '\t':
-                        replaceWith = 't';
-                        break;
-
-                    case '\b':
-                        replaceWith = 'b';
-                        break;
-
-                    case '\v':
-                        replaceWith = 'v';
-                        break;
-
-                    default:
-                        if (quote == c)
-                        {
-                            replaceWith = c;
-                            break;
-                        }
-
-                        if (escapeNonPrintable)
-                        {
-                            switch (CharUnicodeInfo.GetUnicodeCategory(c))
-                            {
-                                case UnicodeCategory.OtherNotAssigned:
-                                case UnicodeCategory.ParagraphSeparator:
-                                case UnicodeCategory.Control:
-                                    unicodeEscape = true;
-                                    break;
-                            }
-                        }
+            Debug.Assert(quote == 0 || quote == '"' || quote == '\'');
 
-                        break;
-                }
-
-                if (unicodeEscape || replaceWith != '\0')
-                {
-                    if (pooledBuilder == null)
+            string replaceWith = null;
+            bool unicodeEscape = false;
+            switch (CharUnicodeInfo.GetUnicodeCategory(c))
+            {
+                case UnicodeCategory.OtherPunctuation:
+                    if (useLanguageSpecificEscapes)
                     {
-                        pooledBuilder = PooledStringBuilder.GetInstance();
-                        sb = pooledBuilder.Builder;
-                        if (quote != 0)
+                        switch (c)
```

Comment:
+1 UnicodeCategories seem to change rather capriciously from version to version...while these specific characters may be pretty *safe*, I'd consider keeping all the common characters as part of the outer switch just in case (as @amcasey suggested).